pkg/balancer/defaultAnalyzer.go

```diff
@@ -47,23 +55,85 @@ func (a *DefaultAnalyzer) AnalyzeLoads() ([]schema.Schedule, bool) {
 	return newSchedules, len(newSchedules) > 0
 }
 
-func (a *DefaultAnalyzer) detectImbalanceShards(loads []*ShardMetrics) ([]uint64, error) {
+func (a *DefaultAnalyzer) detectImbalanceShards(loads []*NodeMetrics) ([]uint64, error) {
 	var imbalancedShards []uint64
+	avgEntries := 0.0
 	for _, load := range loads {
-		if load.CPU > 80.0 || load.Memory > 1<<30 { // 1 GiB
-			imbalancedShards = append(imbalancedShards, load.ShardID)
+		avgEntries += float64(load.NumEntries)	
+	}
+	avgEntries /= float64(len(loads))
+	
+	// If a shard is responsible for more than twice the average number of entries, it is considered imbalanced
+	for idx, load := range loads {
+		if float64(load.NumEntries) > avgEntries * 2 {
+			imbalancedShards = append(imbalancedShards, uint64(idx))
 		}
 	}
 	return imbalancedShards, nil
 }
 
-func (a *DefaultAnalyzer) collectMetrics() ([]*ShardMetrics, error) {
-	// Example: Dummy logic to collect shard metrics
-	return []*ShardMetrics{}, nil
+func queryNodeStats(nodeAddress string) (*NodeMetrics, error) {
+	conn, err := grpc.NewClient(nodeAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	if err != nil {
+		log.Fatalf("Failed to connect to gRPC server: %v", err)
+		return nil, err
+	}
+	defer conn.Close()
+
+	client := pb.NewNodeStatsClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	resp, err := client.GetStats(ctx, &pb.GetStatsRequest{})
+	if err != nil {
+		log.Fatalf("Failed to get stats: %v", err)
+		return nil, err
+	}
+
+	return &NodeMetrics{
+		NumEntries: resp.Entries,
+		NumSuccessfulRequets: resp.SuccessfulRequests,
+		NumFailedRequests: resp.FailedRequests,
+		LastResetTime: resp.LastResetTime.AsTime(),
+	}, nil
+}
+
+func (a *DefaultAnalyzer) collectMetrics() ([]*NodeMetrics, error) {
```

Comment:
Not entirely clear what this method does, presumably this collects metrics for the nodes itself and adds it to the Raft group's log?

Reply:
It collects metrics from all nodes (servers in the replica groups). Metrics are never logged in Raft nor are they synchronized.

Reply:
Note that this is the code for the `Balancer`, which is a part of controller